File: ruby-sdk/lib/cleanster/api/users_api.rb
module Cleanster
  module Api
    # UsersApi - manage end-user accounts and authentication tokens.
    class UsersApi
      def initialize(http)
        @http = http
      end

      # Create a new user account under your partner.
      #
      # Confirmed against the live sandbox API: this requires an undocumented
      # +customer_id+ field, and the response is +{userId, accessToken}+, not a
      # full user profile.
      #
      # @param email       [String] Required.
      # @param first_name  [String] Required.
      # @param last_name   [String] Required.
      # @param customer_id [String] Required. Your own internal customer/user ID.
      # @param phone       [String, nil] Optional.
      # @return [Models::ApiResponse<Models::CreateUserResponse>]
      def create_user(email:, first_name:, last_name:, customer_id:, phone: nil)
        body = { email: email, firstName: first_name, lastName: last_name, customerId: customer_id }
        body[:phone] = phone if phone
        raw = @http.post("/v1/user/account", body: body)
        Models::ApiResponse.from_hash(raw, model_class: Models::CreateUserResponse)
      end

      # Fetch the long-lived bearer token for a user.
      # Pass the returned token to client.access_token=.
      #
      # @param user_id [Integer]
      # @return [Models::ApiResponse<Models::User>]  User with token field populated.
      def fetch_access_token(user_id)
        raw = @http.get("/v1/user/access-token/#{user_id}")
        Models::ApiResponse.from_hash(raw, model_class: Models::User)
      end

      # Verify that a JWT token is valid and has not expired.
      #
      # @param token [String]
      # @return [Models::ApiResponse]
      def verify_jwt(token:)
        raw = @http.post("/v1/user/verify-jwt", body: { token: token })
        Models::ApiResponse.from_hash(raw)
      end
    end
  end
end

File: android-sdk/src/main/kotlin/com/cleanster/android/api/UsersApi.kt
package com.cleanster.android.api

import com.cleanster.android.CleansterError
import com.cleanster.android.model.*
import retrofit2.HttpException
import retrofit2.Retrofit
import retrofit2.http.*
import java.io.IOException

internal interface UsersService {
    @POST("v1/user/account")
    suspend fun createUser(@Body body: CreateUserRequest): ApiResponse<User>

    @GET("v1/user/access-token/{userId}")
    suspend fun fetchAccessToken(@Path("userId") userId: Int): ApiResponse<User>

    @POST("v1/user/verify-jwt")
    suspend fun verifyJwt(@Body body: VerifyJwtRequest): ApiResponse<Any>
}

class UsersApi(retrofit: Retrofit) {
    private val service = retrofit.create(UsersService::class.java)

    suspend fun createUser(
        email: String,
        firstName: String,
        lastName: String,
        phone: String? = null,
    ): ApiResponse<User> = wrap {
        service.createUser(CreateUserRequest(email, firstName, lastName, phone))
    }

    suspend fun fetchAccessToken(userId: Int): ApiResponse<User> = wrap {
        service.fetchAccessToken(userId)
    }

    suspend fun verifyJwt(token: String): ApiResponse<Any> = wrap {
        service.verifyJwt(VerifyJwtRequest(token))
    }
}

internal suspend fun <T> wrap(block: suspend () -> T): T {
    return try {
        block()
    } catch (e: HttpException) {
        if (e.code() == 401) throw CleansterError.Unauthorized()
        throw CleansterError.ApiError(e.code(), e.message())
    } catch (e: IOException) {
        throw CleansterError.NetworkError(e.message ?: "Unknown network error")
    }
}
